### PR TITLE
feat: osmosis multi-hop quote and execution

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

- [x] first and second hops as quote - most of the logic should be moved from trade execution here to properly build a first and second hop depending on ATOM -> OSMO or OSMO -> ATOM direction.
- [x] consume the multi-hop quote within `executeTrade()` and simplify it since most of the logic will have been lifted at quote time. Make sure to do this in a non-breaking way i.e if underlying utils have to be updated, they should be duplicated into a `2` flavor (or perhaps note? Trade quote seems to be happy with the flag off, and so does execution)
  - [x] swap-exact-amount-in
  - [x] IBC transfer
  - [x] ensure this works both directions
- [x] bring back current `getTradeQuote` implementation and make this PR's `getTradeQuote` `getMultiHopTradeQuote` - this is to avoid regressions until multi-hop is fully implemented, and we're able to fully test this end-to-end.
- [x] tackle review comments from https://github.com/shapeshift/web/pull/4903

This PR brings Osmosis swapper as a multi-hop swapper. It is fully implemented both at quote and execution-level, and works with the multi-hop flag on ~~as well as a non-breaking change with the flag off~~

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes https://github.com/shapeshift/web/issues/4837

## Risk

Theoretically none with the flag off.

- Cross account trading may be borked both with the flag on
- Rates are known to be slightly suboptimal here, except a follow-up to improve fees deduction on both hops once we have multi-hop properly implemented

## Testing

- Osmosis swaps are still happy with the multi-hop flag off
- Osmosis multi-account swaps are still happy
- Minimums are still handled
- Fees look sane

### Engineering

Flag off: ☝🏽 

Note, multi-hop execution isn't properly implemented so we'll have to get hacky here:

- First, ensure we're able to get a trade quote with multi-hop on, both directions, then:

1. Execute the trade. This will only execute the first hop. Ensure that you're using a buffer as opposed to the amount you're actually intending on selling, see 2.
2. After the trade is marked as "completed", i.e first hop done, change this guy to `1` i.e second hop:

https://github.com/shapeshift/web/blob/d88fbbf78c402a4619e6bee57331388404ea9d67/src/components/MultiHopTrade/hooks/useTradeExecution/useTradeExecution.ts#L94

Get a trade quote again, and execute the second hop. The buffer mentioned in step 1 will come into play since rate could have changed between the first and the second trade.
Ensure the second hop goes through and is displayed as completed 🎉  this means that once we implement steps in trade execution, this will properly work

Note, `updateFees` doesn't handle multi-hop yet, so trade confirmation amounts will be rugged, as is expected

### Operations

- Multi-hop flag off: ☝🏽 
- Flag on: Nothing to see here yet 👀 

#### ATOM -> OSMO

- stepIndex 0 (IBC transfer)

image

- stepIndex 1 (swap-exact-amount-in)

image

#### OSMO -> ATOM

- stepIndex 0 (swap-exact-amount-in)

image

- stepIndex 1 (IBC transfer)

image

#### ATOM -> OSMO

image

#### OSMO -> ATOM

image

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #4901** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4901" target="_blank">Graphite</a> 
    * **PR #4903** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4903" target="_blank">Graphite</a> 
      * **PR #4905** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4905" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/4905?utm_source=stack-comment).

File: src/lib/swapper/swappers/OsmosisSwapper/endpoints.ts [hunk 1]
```diff
@@ -60,75 +54,79 @@ export const osmosisApi: Swapper2Api = {
     tradeQuote,
     stepIndex,
   }: GetUnsignedTxArgs): Promise<CosmosSignTx> => {
+    if (!from) throw new Error('from address is required')
```
Comment (woodenfurniture) [inline src/lib/swapper/swappers/OsmosisSwapper/endpoints.ts:57]: ❤️

File: src/lib/swapper/swappers/OsmosisSwapper/endpoints.ts [hunk 2]
```diff
@@ -60,75 +54,79 @@ export const osmosisApi: Swapper2Api = {
...
 
       const unsignedTx = await buildPerformIbcTransferUnsignedTx({
         input: transfer,
-        adapter: cosmosAdapter,
-        blockBaseUrl: osmoUrl,
-        denom: 'uatom',
-        sourceChannel: COSMO_OSMO_CHANNEL,
-        feeAmount: ibcFromCosmosFeeData.fast.txFee,
+        adapter: sellAssetAdapter,
+        // The blockheight of the destination chain for the IBC transfer
```
Comment (woodenfurniture) [inline src/lib/swapper/swappers/OsmosisSwapper/endpoints.ts:105]: this comment seems out of place 😅

Comment (gomesalexandre) [reply src/lib/swapper/swappers/OsmosisSwapper/endpoints.ts]: eyy nice catch, needs a little rewordy cea29eec3f98b84b68094d94294411085ec8da27

File: src/lib/swapper/swappers/OsmosisSwapper/getTradeQuote/getTradeQuote.ts [hunk 3]
```diff
@@ -51,42 +59,122 @@ export const getTradeQuote = async (
...
+    allowanceContract: '',
+    buyAsset: firstHopBuyAsset,
+    feeData: {
+      networkFeeCryptoBaseUnit: firstHopNetworkFee,
+      protocolFees: {
+        ...(buyAssetIsOnOsmosisNetwork
+          ? {}
+          : {
+              [sellAsset.assetId]: {
+                // TODO(gomes): is this correct? There may be "0 fees" as amounts from rates are pessimistic by nature and assume fees
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/OsmosisSwapper/getTradeQuote/getTradeQuote.ts:129]: Investigate before opening me, we might occur protocol fees that shouldn't here

File: src/lib/swapper/swappers/OsmosisSwapper/endpoints.ts [hunk 4]
```diff
@@ -45,12 +40,7 @@ export const osmosisApi: Swapper2Api = {
     return tradeQuoteResult.map(tradeQuote => {
       const { receiveAddress, affiliateBps } = input
       const id = uuid()
-      // TODO(gomes): getTradeQuote() is currently a hack, which represents the swap as a single trade, and needs to be revamped
-      // This effectively means that getUnsignedTx() won't be able to be implemented for all steps, but only the first for now
-      // i.e either the IBC transfer or swap-exact-amount-in
-      tradeQuoteMetadata.set(id, {
-        chainId: tradeQuote.steps[0].sellAsset.chainId as OsmosisSupportedChainId,
-      })
+      tradeQuoteMetadata.set(id, input as GetCosmosSdkTradeQuoteInput)
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/OsmosisSwapper/endpoints.ts:43]: Not to have to fiddle with first/last hop things but maybe that's overkill

File: src/lib/swapper/swappers/OsmosisSwapper/getTradeQuote/getMultiHopTradeQuote.ts [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+import { assertGetCosmosSdkChainAdapter } from 'lib/utils/cosmosSdk'
+
+import { DEFAULT_SOURCE } from '../utils/constants'
+
+export const getTradeQuote = async (
+  input: GetTradeQuoteInput,
+  { sellAssetUsdRate }: { sellAssetUsdRate: string },
+): Promise<Result<TradeQuote<ChainId>, SwapErrorRight>> => {
+  const {
+    // TODO(gomes): very very dangerous. We currently use account number both on the sending and receiving side and this will break cross-account.
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/lib/swapper/swappers/OsmosisSwapper/getTradeQuote/getMultiHopTradeQuote.ts:31]: Actually can't properly be tested until multi-hop execution supports executing all steps

File: src/state/slices/tradeQuoteSlice/helpers.ts [hunk 6]
```diff
@@ -165,9 +188,12 @@ export const getTotalProtocolFeeByAsset = (quote: TradeQuote2): Record<AssetId,
...
 
-        innerAcc[assetId].amountCryptoBaseUnit = bn(innerAcc[assetId].amountCryptoBaseUnit)
-          .plus(protocolFee.amountCryptoBaseUnit)
-          .toString()
+        innerAcc[assetId] = {
+          ...innerAcc[assetId],
+          amountCryptoBaseUnit: bn(innerAcc[assetId].amountCryptoBaseUnit)
+            .plus(protocolFee.amountCryptoBaseUnit)
+            .toString(),
+        }
```
Comment (gomesalexandre) [inline src/state/slices/tradeQuoteSlice/helpers.ts:196]: Was getting turbo weird frozen/read-only object errors here

```shell
Uncaught TypeError: Cannot assign to read only property 'amountCryptoBaseUnit' of object '#<Object>'
The error may be correlated with this previous error:
TypeError: Cannot assign to read only property 'amountCryptoBaseUnit' of object '#<Object>'
    at eval (webpack-internal:///./src/state/slices/tradeQuoteSlice/helpers.ts:140:44)
    at Array.reduce (<anonymous>)
    at eval (webpack-internal:///./src/state/slices/tradeQuoteSlice/helpers.ts:132:52)
    at Array.reduce (<anonymous>)
    at getTotalProtocolFeeByAsset (webpack-internal:///./src/state/slices/tradeQuoteSlice/helpers.ts:131:57)
    at eval (webpack-internal:///./src/state/slices/tradeQuoteSlice/selectors.ts:127:255)
    at recomputationWrapper (webpack-internal:///./node_modules/reselect/es/index.js:76:25)
    at memoized (webpack-internal:///./node_modules/reselect/es/defaultMemoize.js:129:20)
    at dependenciesChecker (webpack-internal:///./node_modules/reselect/es/index.js:90:40)
    at Object.memoized [as current] (webpack-internal:///./node_modules/reselect/es/defaultMemoize.js:129:20)
```

**Review by 0xApotheosis — APPROVED** [review]
I wan't able to test the Osmosis swapper as the node is currently down, though the code looks good and I noticed no regressions to the production swapper.

I did notice some console spew when the multi-hop flag was turned on relating to pools, but we can tackle that in a separate PR.

**Pull request merged**